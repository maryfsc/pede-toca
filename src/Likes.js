import React from 'react';
import { Button, Badge } from 'reactstrap';

class Likes extends React.Component {
  
  constructor(props) {
    super(props);
    this.state = {
      likes: 0
    }
  }

  _handleLikes() {    
    this.setState({ likes: this.state.likes + 1 });
  };

  render() {
    return (
      <div className="like-wrapper">
        <Button color="secondary" onClick={() => this._handleLikes()}> 
          Like<Badge color="secondary">{this.state.likes}</Badge>
        </Button>
      </div>
    )
  }
}

export default Likes;

/*
1) What it renders?
  -a button with a badge (span) that contains Likes and a counter for them, starting with 0;

2) Does the component renders different things under different circumstances?
  -No, it's always a button with a badge.

3) When passing a function as a prop, does it calls it? What does it does?
  -Yes, it should make a callback call to update the state;

4) What happens when the user interacts with my component?
  -When the user clicks it, it should update the like count accordingly.

Other contracts:
  - a div is always rendered, and it contains all other elements;
  - there is a button and inside of it, a span with like counts;?
  */
