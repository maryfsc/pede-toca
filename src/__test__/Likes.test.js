import React from 'react';
import { shallow } from 'enzyme';
import Likes from '../Likes';

describe('<Likes />', () => {
  const wrapper = shallow(<Likes />);

  it('renders correctly', () => {

    expect(wrapper.find(<div className="likes-wrapper"></div>)).toBeDefined();
    expect(wrapper.find(<span className="badge badge-secondary" />)).toBeDefined(); 
    // expect(wrapper.find(<button className="btn btn-secondary" />).contains('Like')).toBeTruthy(); // ????? 
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  // it('updates count likes when the button is clicked', () => {
  //   const button = wrapper.find(<Button />);
  //   const badge = wrapper.findWhere(node => node.InnerText === '0');

  //   button.simulate('click'); // error: Method “simulate” is meant to be run on 1 node. 0 found instead.

  //   expect(badge.contains(1)).toBeTruthy();
  // });
});

/*
Summary of React Tests:    
  Find your Component Contract first
  Decide which constraints are worth testing and which aren’t
  Prop types are not worth testing
  Inline styles are usually not worth testing
  The components you render and what props you give them are important to test
  Don’t test things that are not the concern of your component
*/

/* 
  5) Como testar funções callback?
  6) Como mockar uma função? Tem um método do jest (jest.fn() ?) pra isso pelo que entendi, mas não entendi como funciona

  To Do:
    - estudar como mockar com jest (done);
      -shallow do Enzyme é como se fosse um mock?
    
    - diferenças entre mock x spy (done):
      -Mock é possível de ser construído 'do nada', criando uma classe inteira e métodos para teste. 
      -Spy é usado pra selecionar geralmente uma parte da classe que será testada e utiliza métodos REAIS, "espiando" o objeto real em questão; 
    
    - como testar o chamamento de uma função;
    
    - mudar [?] a lib pra alguma que atenda as necessidades da aplicação (enzyme?)(done);
      -escolhi enzyme;
*/