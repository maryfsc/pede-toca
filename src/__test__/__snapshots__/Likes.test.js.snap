// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Likes /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Likes />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Like
        <Badge
          color="secondary"
          pill={false}
          tag="span"
        >
          0
        </Badge>
      </Button>,
      "className": "like-wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          "Like",
          <Badge
            color="secondary"
            pill={false}
            tag="span"
          >
            0
          </Badge>,
        ],
        "color": "secondary",
        "onClick": [Function],
        "tag": "button",
      },
      "ref": null,
      "rendered": Array [
        "Like",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": 0,
            "color": "secondary",
            "pill": false,
            "tag": "span",
          },
          "ref": null,
          "rendered": 0,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Button
          color="secondary"
          onClick={[Function]}
          tag="button"
        >
          Like
          <Badge
            color="secondary"
            pill={false}
            tag="span"
          >
            0
          </Badge>
        </Button>,
        "className": "like-wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Like",
            <Badge
              color="secondary"
              pill={false}
              tag="span"
            >
              0
            </Badge>,
          ],
          "color": "secondary",
          "onClick": [Function],
          "tag": "button",
        },
        "ref": null,
        "rendered": Array [
          "Like",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": 0,
              "color": "secondary",
              "pill": false,
              "tag": "span",
            },
            "ref": null,
            "rendered": 0,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
